src/wasm-lib/kcl/src/walk/ast_visitor.rs

```diff
@@ -0,0 +1,197 @@
+    /// should only contain direct descendants.
+    fn children(&self) -> Vec<Node<'tree>>;
+
+    /// Return `self` as a [Node]. Generally speaking, the [Visitable] trait
+    /// is only going to be implemented on [Node], so this is purely used by
+    /// helpers that are generic over a [Visitable] and want to deref back
+    /// into a [Node].
+    fn node(&self) -> Node<'tree>;
+
+    /// Call the provided [Visitor] in order to Visit `self`. This will
+    /// only be called on `self` -- the [Visitor] is responsible for
+    /// recursing into any children, if desired.
+    fn visit<VisitorT>(&self, visitor: VisitorT) -> Result<bool, VisitorT::Error>
+    where
+        VisitorT: Visitor<'tree>,
+    {
+        visitor.visit_node(self.node())
+    }
+}
+
+/// Trait used to enable visiting members of KCL AST.
+///
+/// Implementing this trait enables the implementor to be invoked over
+/// members of KCL AST by using the [Visitable::visit] function on
+/// a [Node].
+pub trait Visitor<'tree> {
+    /// Error type returned by the [Self::visit] function.
+    type Error;
+
+    /// Visit a KCL AST [Node].
+    ///
+    /// In general, implementers likely wish to check to see if a Node is what
+    /// they're looking for, and either decend into that [Node]'s children (by
+    /// calling [Visitable::children] on [Node] to get children nodes,
+    /// calling [Visitable::visit] on each node of interest), or preform
+    /// some action.
+    fn visit_node(&self, node: Node<'tree>) -> Result<bool, Self::Error>;
+}
+
+impl<'a, FnT, ErrorT> Visitor<'a> for FnT
+where
+    FnT: Fn(Node<'a>) -> Result<bool, ErrorT>,
+{
+    type Error = ErrorT;
+
+    fn visit_node(&self, n: Node<'a>) -> Result<bool, ErrorT> {
+        self(n)
+    }
+}
+
+impl<'tree> Visitable<'tree> for Node<'tree> {
+    fn node(&self) -> Node<'tree> {
+        *self
+    }
+
+    fn children(&self) -> Vec<Node<'tree>> {
+        match self {
+            Node::Program(n) => n.body.iter().map(|node| node.into()).collect(),
+            Node::ExpressionStatement(n) => {
+                vec![(&n.expression).into()]
+            }
+            Node::BinaryExpression(n) => {
+                vec![(&n.left).into(), (&n.right).into()]
+            }
+            Node::FunctionExpression(n) => {
+                let mut children = n.params.iter().map(|v| v.into()).collect::<Vec<Node>>();
+                children.push((&n.body).into());
+                children
+            }
+            Node::CallExpression(n) => {
+                let mut children = n.arguments.iter().map(|v| v.into()).collect::<Vec<Node>>();
+                children.insert(0, (&n.callee).into());
+                children
+            }
+            Node::CallExpressionKw(n) => {
+                let mut children = n.unlabeled.iter().map(|v| v.into()).collect::<Vec<Node>>();
+
+                // TODO: this is wrong but it's what the old walk code was doing.
```

Comment:
I need to re-read this again now that the PR is functional to see if this needs any more work, likely maybe